### ISSUE TITLE
How to convert image.Image to Mat?

### ISSUE THREAD

**Issue by edwvee** [hybridgroup/gocv#228]
Need to use FindContours. It receives Mat as source. Searched project through for "image.Image". Found only encoding Mat to image.Image. Found NewMatFromBytes. But how should I encode image.Image's data in []byte? Stuck on this code:
```go
bounds := img.Bounds()
x := bounds.Dx()
y := bounds.Dy()
//getting []bytes
//???
gocv.NewMatFromBytes(x, y, myBytes)
```
Does it appropriate to ask help via issues? Haven't found gocv forum or etc.

**Comment by rikonor** [comment]
You need to encode your image to a format that OpenCV can read, like jpeg, png, etc.

Try feeding `NewMatFromBytes` the original file data or encode your `img` into a buffer as one of those formats and give that to `NewMatFromBytes`.

**Comment by edwvee** [comment]
Is there no other way? That's not a good option, because encoding/decoding is much expensive. If it's an only option what is the reason use go instead of cpp for opencv?

**Comment by rikonor** [comment]
This library provides bindings, not an implementation, so you have to provide it input that it understands.

You might have better luck doing your initial processing in OpenCV instead of Go's native image library, that way you don't need to do an additional conversion.

Another option would be to convert your `image.Image` to `ppm` format which would be somewhat lower cost then the other formats, then providing that to OpenCV since it should understand that format.

**Comment by edwvee** [comment]
Well, have tried to get pixels from image and pass them as bytes. It works, but there are troubles with colors.
How to convert 32bit uint values from RGBA() (r, g, b, a uint32) to 8bit? Something like ```byte(r >> 8)``` gives a result, but not mostly correct. Haven't yet understood the concept of 32bit components.

**Comment by edwvee** [comment]
Yes! That is!
```go
bytes := make([]byte, 0, x*y)
	for j := bounds.Min.Y; j < bounds.Max.Y; j++ {
		for i := bounds.Min.X; i < bounds.Max.X; i++ {
			r, g, b, _ := img.At(i, j).RGBA()
			bytes = append(bytes, byte(b>>8))
			bytes = append(bytes, byte(g>>8))
			bytes = append(bytes, byte(r>>8))
		}
	}

	mat, err := gocv.NewMatFromBytes(y, x, gocv.MatTypeCV8UC3, bytes)
```

**Comment by deadprogram** [comment]
Hi, @edwvee this seems like a useful addition, perhaps you might submit a PR a `gocv.NewMatFromImage()` function to `core.go`?

One suggestion, you can handle the `A` param if you change to this:

```go
r, g, b, a := img.At(i, j).RGBA()
bytes = append(bytes, byte(b>>8))
bytes = append(bytes, byte(g>>8))
bytes = append(bytes, byte(r>>8))
bytes = append(bytes, byte(a>>8))
```

and also change this:

```go
mat, err := gocv.NewMatFromBytes(y, x, gocv.MatTypeCV8UC4, bytes)
```

Making it a 4 channel `Mat` able to accommodate the alpha channel.

What do you think?

**Comment by edwvee** [comment]
However got no clue how to pass encoded format to gocv for receiving proper Mat. Looks like there is no method allowing that. Only creating file and passing to ImRead(or how the function called).

**Comment by edwvee** [comment]
Bmp from here works: https://godoc.org/golang.org/x/image/bmp . But have the same performance.
From opencv documentation:

> Currently, the following file formats are supported:
> Windows bitmaps - *.bmp, *.dib (always supported)
> JPEG files - *.jpeg, *.jpg, *.jpe (see the Notes section)
> JPEG 2000 files - *.jp2 (see the Notes section)
> Portable Network Graphics - *.png (see the Notes section)
> WebP - *.webp (see the Notes section)
> Portable image format - *.pbm, *.pgm, *.ppm *.pxm, *.pnm (always supported)
> Sun rasters - *.sr, *.ras (always supported)
> TIFF files - *.tiff, *.tif (see the Notes section)
> OpenEXR Image files - *.exr (see the Notes section)
> Radiance HDR - *.hdr, *.pic (always supported)
> Raster and Vector geospatial data supported by Gdal (see the Notes section)

Tried ppm, as suggested above, but it didn't work (https://github.com/lmittmann/ppm). There was error "ppm: color mode not supported".  What of enlisted here can be faster than bmp?
So, maybe my algorithm is not that bad. I'm about to make PR. Any suggestions?

**Comment by deadprogram** [comment]
This code is now released as part of v0.14.0 so closing. Thank you!

**Comment by cedricve** [comment]
> This code is now released as part of v0.14.0 so closing. Thank you!

is it in 0.26 ? Can't find it. @deadprogram

**Comment by deadprogram** [comment]
Please see: 
https://pkg.go.dev/gocv.io/x/gocv#ImageGrayToMatGray
https://pkg.go.dev/gocv.io/x/gocv#ImageToMatRGB
https://pkg.go.dev/gocv.io/x/gocv#ImageToMatRGBA
